path_login.go

```diff
@@ -0,0 +1,237 @@
+			Audience: jwt.Audience(role.BoundAudiences),
+			Time:     time.Now(),
+		}
+
+		if err := claims.Validate(expected); err != nil {
+			return logical.ErrorResponse(errwrap.Wrapf("error validating claims: {{err}}", err).Error()), nil
+		}
+
+	case config.OIDCIssuerURL != "":
+		provider, err := b.getProvider(ctx, config)
+		if err != nil {
+			return nil, errwrap.Wrapf("error getting provider for login operation: {{err}}", err)
+		}
+
+		verifier := provider.Verifier(&oidc.Config{
+			SkipClientIDCheck: true,
+		})
+
+		idToken, err := verifier.Verify(ctx, token)
+		if err != nil {
+			return logical.ErrorResponse(errwrap.Wrapf("error validating signature: {{err}}", err).Error()), nil
+		}
+
+		if err := idToken.Claims(allClaims); err != nil {
+			return logical.ErrorResponse(errwrap.Wrapf("unable to successfully parse all claims from token: {{err}}", err).Error()), nil
+		}
+
+		if role.BoundSubject != "" && role.BoundSubject != idToken.Subject {
+			return logical.ErrorResponse("sub claim does not match bound subject"), nil
+		}
+		if len(role.BoundAudiences) != 0 {
+			var found bool
+			for _, v := range role.BoundAudiences {
+				if strutil.StrListContains(idToken.Audience, v) {
+					found = true
+					break
+				}
+			}
+			if !found {
+				return logical.ErrorResponse("aud claim does not match any bound audience"), nil
+			}
+		}
+
+	default:
+		return nil, errors.New("unhandled case during login")
+	}
+
+	userClaimRaw, ok := allClaims[role.UserClaim]
+	if !ok {
+		return logical.ErrorResponse(fmt.Sprintf("%q claim not found in token", role.UserClaim)), nil
+	}
+	userName, ok := userClaimRaw.(string)
+	if !ok {
+		return logical.ErrorResponse(fmt.Sprintf("%q claim could not be converted to string", role.UserClaim)), nil
+	}
+
+	var groupAliases []*logical.Alias
+	if role.GroupsClaim != "" {
+		groupsClaimRaw, ok := allClaims[role.GroupsClaim]
+		if !ok {
+			return logical.ErrorResponse(fmt.Sprintf("%q claim not found in token", role.GroupsClaim)), nil
+		}
+		groups, ok := groupsClaimRaw.([]interface{})
+		if !ok {
+			return logical.ErrorResponse(fmt.Sprintf("%q claim could not be converted to string list", role.GroupsClaim)), nil
+		}
+		for _, groupRaw := range groups {
+			group, ok := groupRaw.(string)
+			if !ok {
+				return logical.ErrorResponse(fmt.Sprintf("value %v in groups claim could not be parsed as string", groupRaw)), nil
+			}
+			if group == "" {
+				continue
+			}
+			groupAliases = append(groupAliases, &logical.Alias{
+				Name: group,
+			})
+		}
+	}
+
+	resp := &logical.Response{
+		Auth: &logical.Auth{
+			Policies:    role.Policies,
+			DisplayName: userName,
+			Period:      role.Period,
+			NumUses:     role.NumUses,
+			Alias: &logical.Alias{
+				Name: userName,
+			},
+			GroupAliases: groupAliases,
+			InternalData: map[string]interface{}{
+				"role": roleName,
+			},
+			Metadata: map[string]string{
+				"role": roleName,
+			},
+			LeaseOptions: logical.LeaseOptions{
+				Renewable: true,
```

Comment:
Were you going to make this non-renewable unless it's periodic?

Reply:
Following the azure/kube backends, I decided to just make it renewable; you can always set the max TTL to the same as the TTL.